app/app.js

```diff
@@ -0,0 +1,190 @@
+const express = require('express')
+const app = express()
+const bodyParser = require('body-parser')
+const nunjucks = require('nunjucks')
+const util = require('util')
+const fs = require('fs')
+const path = require('path')
+
+const readdir = util.promisify(fs.readdir)
+
+const helperFunctions = require('../lib/helper-functions')
+const fileHelper = require('../lib/file-helper')
+const configPaths = require('../config/paths.json')
+
+// Set up views
+const appViews = [
+  configPaths.layouts,
+  configPaths.views,
+  configPaths.examples,
+  configPaths.fullPageExamples,
+  configPaths.components,
+  configPaths.src,
+  configPaths.node_modules
+]
+
+module.exports = (options) => {
+  const nunjucksOptions = options ? options.nunjucks : {}
+
+  // Configure nunjucks
+  const env = nunjucks.configure(appViews, {
+    autoescape: true, // output with dangerous characters are escaped automatically
+    express: app, // the express app that nunjucks should install to
+    noCache: true, // never use a cache and recompile templates each time
+    trimBlocks: true, // automatically remove trailing newlines from a block/tag
+    lstripBlocks: true, // automatically remove leading whitespace from a block/tag
+    watch: true, // reload templates when they are changed. needs chokidar dependency to be installed
+    ...nunjucksOptions // merge any additional options and overwrite defaults above.
+  })
+
+  // make the function available as a filter for all templates
+  env.addFilter('componentNameToMacroName', helperFunctions.componentNameToMacroName)
+
+  // Add mock `url_for` for components that get urls from Flask
+  env.addGlobal('url_for', function () { return '#' })
+
+  // Set view engine
+  app.set('view engine', 'njk')
+
+  // Disallow search index indexing
+  app.use(function (req, res, next) {
+    // none - Equivalent to noindex, nofollow
+    // noindex - Do not show this page in search results and do not show a
+    //   "Cached" link in search results.
+    // nofollow - Do not follow the links on this page
+    res.setHeader('X-Robots-Tag', 'none')
+    next()
+  })
+
+  // Set up middleware to serve static assets
+  app.use('/public', express.static('app/public/'))
+
+  app.use('/docs', express.static(configPaths.sassdoc))
+
+  // serve html5-shiv from node modules
+  app.use('/vendor/html5-shiv/', express.static('node_modules/html5shiv/dist/'))
+
+  // serve legacy code from node node modules
+  app.use('/vendor/govuk_template/', express.static('node_modules/govuk_template_jinja/assets/'))
+
+  app.use('/assets', express.static(path.join(configPaths.src, 'assets')))
+
+  // Turn form POSTs into data that can be used for validation.
+  app.use(bodyParser.urlencoded({ extended: true }))
+
+  // Handle the banner component serverside.
+  require('./banner.js')(app)
+
+  // Define routes
+
+  // Index page - render the component list template
+  app.get('/', async function (req, res) {
```

Comment:
Looking at the design system, these route declarations usually live in a `routes.js`. I quite like that separation. But I don't know if that's an Express convention or a Design System convention

Reply:
I agree that's a nice convention, I also think though that we should try and keep the app javascript as close to what GOV.UK Frontend uses, so we can keep up with any features they add/bugs they fix in future...

Reply:
Just some background info, Express.js is not opinionated which is why there isn't really a set convention if you try to find one. A lot of patterns are copied from other MVC type frameworks.

Reply:
👍 Fair enough